sphinxext/opengraph/socialcards.py

```diff
@@ -0,0 +1,237 @@
+
+    # Image axis
+    ax_x, ax_y, ax_w, ax_h = (0.69, 0.7, 0.25, 0.25)
+    axim_logo = fig.add_axes((ax_x, ax_y, ax_w, ax_h), anchor="NE")
+
+    # Image shadow axis
+    ax_x, ax_y, ax_w, ax_h = (0.82, 0.1, 0.1, 0.1)
+    axim_shadow = fig.add_axes((ax_x, ax_y, ax_w, ax_h), anchor="NE")
+
+    # Line at the bottom axis
+    axline = fig.add_axes((-0.1, -0.04, 1.2, 0.1))
+
+    # Axes configuration
+    left_margin = 0.05
+    with plt.rc_context({"font.family": font.name, "text.color": text_color}):
+        # Site title
+        # Smaller font, just above page title
+        site_title_y_offset = 0.9
+        txt_site = axtext.text(
+            left_margin,
+            site_title_y_offset,
+            "Test site title",
+            {
+                "size": 26,
+            },
+            ha="left",
+            va="top",
+            wrap=True,
+        )
+
+        # Page title
+        # A larger font for more visibility
+        page_title_y_offset = 0.8
+
+        txt_page = axtext.text(
+            left_margin,
+            page_title_y_offset,
+            "Test page title, a bit longer to demo",
+            {"size": 42, "color": "k", "fontweight": "bold"},
+            ha="left",
+            va="top",
+            wrap=True,
+        )
+
+        txt_page._get_wrap_line_width = _set_pagetitle_line_width
+
+        # description
+        # Just below site title, smallest font and many lines.
+        # Our target length is 160 characters, so it should be
+        # two lines at full width with some room to spare at this length.
+        description_y_offset = 0.22
+        txt_description = axtext.text(
+            left_margin,
+            description_y_offset,
+            (
+                "A longer description that we use to ,"
+                "show off what the descriptions look like."
+            ),
+            {"size": 16},
+            ha="left",
+            va="bottom",
+            wrap=True,
+        )
+        txt_description._get_wrap_line_width = _set_description_line_width
+
+        # url
+        # Aligned to the left of the shadow image
+        url_y_axis_ofset = 0.12
+        txt_url = axtext.text(
+            left_margin,
+            url_y_axis_ofset,
+            "testurl.org",
+            {"size": 22},
+            ha="left",
+            va="bottom",
+            fontweight="bold",
+        )
+
+    if image_shadow:
+        img = mpimg.imread(image_shadow)
+        axim_shadow.imshow(img)
+
+    # Put the logo in the top right if it exists
+    if image:
+        img = mpimg.imread(image)
+        axim_logo.imshow(img)
+
+    # Put a colored line at the bottom of the figure
+    axline.hlines(0, 0, 1, lw=16, color=line_color)
+
+    # Remove the ticks and borders from all axes for a clean look
+    for ax in fig.axes:
+        ax.set_axis_off()
+    return fig, txt_site, txt_page, txt_description, txt_url
+
+
+def render_social_card(app, sitetitle, pagetitle, description, siteurl, pagepath):
```

Comment:
I think that `render_social_card` should be given a path to write an image to and not return anything. `def render_social_card(out_path, **kwargs) -> None` for example. This way, people who want to have custom image generation only have to provide one function.

Reply:
hmm - ok, I can give this a shot but I think it will mean refactoring a large chunk of this functionality. I'll give it a shot though.